---
title: "SBF vignette"
author: "Amal Thomas"
output: rmarkdown::html_vignette
bibliography: references.bib
vignette: >
  %\VignetteIndexEntry{SBF_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style>
body {
text-align: justify}
</style>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


## Background

Joint matrix factorization facilitates the comparison of
expression profiles from different species without using gene mapping.
Transforming gene expression profiles into reduced eigengene space using
singular value decomposition (SVD) has been shown to capture meaningful
biological information [@alter2000singular].
@tamayo2007metagene used a non-negative matrix factorization
approach to learn a low-dimensional approximation of the
microarray expression datasets and used the reduced space for comparisons.
Matrix factorization based methods are commonly used for gene expression
analysis [@alter2000singular; @tamayo2007metagene].
An orthology independent matrix factorization framework based on generalized
singular value decomposition [GSVD; @van1976generalizing] was used
by @alter2003generalized to compare gene-expression
profiles from two species.
This framework was later extended to develop higher-order generalized singular
value decomposition (HO GSVD) to analyze data from more than two species 
[@ponnapalli2011higher].

This study developed a joint diagonalization approach called
approximate shared basis factorization (A-SBF)
for cross-species expression comparisons.
This approach extends the exact factorization approach we 
developed called shared basis factorization (SBF). We discuss
the details of the two approaches in the following sections.


## Shared basis factorization

Consider a set of real matrices $D_i \in \mathbb{R}^{m_i \times k}$
($i={1,\ldots ,N}$) with full column rank. We define 
shared basis factorization (SBF) as

\begin{align*}
  D_1 &= U_1\Delta_1V^T, \\
  D_2 &= U_2\Delta_2V^T, \\
      & \vdots \\
  D_N &= U_N\Delta_NV^T.
\end{align*}

Here each $U_i \in \mathbb{R}^{m_i \times k}$ is a dataset-specific
left basis matrix, each $\Delta_i \in \mathbb{R}^{k \times k}$ is a
diagonal matrix with positive values $\delta_{ik}$, and $V$ is
a square invertible matrix.

### Learning the shared right basis matrix

Let $M$ be the scaled sum of the $D_i^T D_i$.
We define $M$ is defined as
\[
 M = \frac{\sum_{i=1}^{N} D_i^T D_i/w_i}{\alpha}.
\]

The scaling factor $w_i$ is the total variance explained by the column
vectors of $D_i$, and $\alpha$ is the inverse sum of the total variance of
$D_i$, for $i = 1 \cdots N$. The weights $w_i$ and $\alpha$ are defined as

\begin{align*}
    w_i &= \sum_{j=1}^{k} \sigma_{jj}^{2\mbox{ }(i)} \mbox{ and}\\
    \alpha &= \sum_{i=1}^{N} \frac{1}{\sum_{j=1}^{k} \sigma_{jj}^{2\mbox{ }(i)}}.
\end{align*}

Here $\sum_{j=1}^{k} \sigma_{jj}^{2\mbox{ }(i)} = \tr(D_i^T D_i) = \tr(A_i)$.
Using the $w_i$ and $\alpha$, individual $D_i^T D_i$ are standardized.
If all the variances are equal, $M$ becomes the arithmetic mean of the sum of
$D_i^T D_i$.
The shared right basis matrix  $V$ is then determined from the eigenvalue
decomposition of $M$, where $M=V \Theta V^T$.
Given $V$, we compute $U_i$ and $\Delta_i$ by solving the linear system
$D_i V = U_i \Delta_i = L_i$.
By normalizing the columns of $L_i$, we have
$\delta_{ik} = \|l_{ik}\|$ and
$\Delta_i = \mbox{diag}(\delta_{i1},\ldots,\delta_{ik})$.

## Approximate shared basis factorization

Consider a set of matrices $D_i \in \mathbb{R}^{m_i \times k}$
($i= 1,\ldots,N$), each with full column rank. We define
approximate shared basis factorization (A-SBF) as

\begin{align*}
  D_1 &= U_1\Delta_1V^T + \epsilon_1, \\
  D_2 &= U_2\Delta_2V^T + \epsilon_2, \\
      & \vdots \\
  D_N &= U_N\Delta_NV^T + \epsilon_N.
\end{align*}

Each $U_i \in \mathbb{R}^{m_i \times k}$ is a species-specific left basis
matrix with **orthonormal** columns (eigengenes),
$\Delta_i \in \mathbb{R}^{k \times k}$ is a diagonal matrix with positive
values $\Delta_{ik}$ and $V$ is a non singular square matrix.
The right basis matrix $V$ is identical in all the $N$ matrix factorizations
and defines the common space shared by all species. 
We learn the factorization such that the learned $V$ is closest to that in
the exact decomposition and by minimizing the total decomposition error 
$\sum^{N}_{i=1}\epsilon_i = {\sum^{N}_{i=1}\|D_i - U_i\Delta_iV^T\|^2}_F$.

```{r setup}
# load SBF package
library(SBF)
```

Lets load SBF package's in-built gene expression dataset. The dataset contains
average gene expression profile of five similar tissues in three species.
```{r}
# load dataset
avg_counts <- SBF::TissueExprSpecies
# check the names of species
names(avg_counts)
```

```{r}
# head for first species
avg_counts[[names(avg_counts)[1]]][1:5,1:5]
```

The number of genes annotated in different species are different. As a result,
the number of rows (genes) in the expression data will be different for
different species.

```{r}
sapply(avg_counts, dim)
```

## References
